docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,374 @@
+
+The plugin-ware verifier application, which can either be the enclave or the host, can verify the evidence like this:
+
+`verifier.c`
+
+```C
+#include <sgx/verifier.h>
+
+/* Register SGX DCAP plugin upon initialization. */
+oe_register_verifier(oe_sgx_plugin_dcap_verifier(), NULL, 0);
+
+/* Get registered verifier format IDs. */
+oe_get_registered_verifier_format_ids(&format_ids, &format_ids_length);
+
+/* Send the format IDs that can be verified to the attester.
+   Protocol is up to enclave and verifier. */
+send(VERIFIER_SOCKET_FD, format_ids, format_ids_length*sizeof(oe_uuid_t), 0);
+
+/* Receive a selected format ID. */
+recv(ENCLAVE_SOCKET_FD, &format_id, &format_id_size, 0);
+
+/* Get claims for the format ID. */
+oe_get_verifier_claims(format_id, &verifier_claims, &verifier_claims_length);
+
+/* Serialize the claims and send the claims buffer to the attester */
+_serialize_claims(verifier_claims, verifier_claims_length, &claims_buffer, &claims_buffer_size);
+send(VERIFIER_SOCKET_FD, claims_buffer, claims_buffer_size, 0);
+
+/* Receive evidence and endorsement buffer from enclave. */
+recv(ENCLAVE_SOCKET_FD, evidence, evidence_size, 0);
+recv(ENCLAVE_SOCKET_FD, endorsements, endorsements_size, 0);
+
+/* Verify evidence. Can check the claims if desired. */
+oe_verify_evidence(
+    evidence,
+    evidence_size,
+    endorsements,
+    endorsements_size,
+    NULL,
+    0,
+    &claims,
+    &claims_size);
+
+/* Free data and unregister plugin. */
+oe_free_claims_list(verifier_claims, verifier_claims_length);
+oe_free_format_ids(format_ids);
+oe_free_claims_list(claims, claims_size);
+oe_unregister_verifier(oe_sgx_plugin_dcap_verifier());
+```
+
+The plugin-aware application enclave can generate the SGX evidence using the plugin like this:
+
+`attester.c`
+
+```C
+#include <sgx/attester.h>
+
+/* Register SGX plugins upon initialization. */
+oe_register_attester(oe_sgx_plugin_dcap_attester(), NULL, 0);
+oe_register_attester(oe_sgx_plugin_local_attester(), NULL, 0);
+
+/* Receive format IDs from verifier. Protocol is up to enclave and verifier. */
+recv(ENCLAVE_SOCKET_FD, &format_ids, &format_ids_size, 0);
+
+/* Two attester plugins were registered, but the verifier can only
+   verify evidence generated by SGX DCAP attester. The selected_format will be
+   SGX DCAP which application simply passes in to oe_get_evidence(). */
+oe_select_attester_format(format_ids, format_ids_size/sizeof(oe_uuid_t), &selected_format);
+
+/* Send the selected format to attester */
+send(ENCLAVE_SOCKET_FD, selected_format, sizeof(oe_uuid_t), 0);
+
+/* Receive claims and de-serialize the buffer. */
+recv(ENCLAVE_SOCKET_FD, &claims_buffer, &claims_buffer_size, 0);
+_deserialize_claims(claims_buffer, claims_buffer_size, &claims, &claims_length);
```

Comment:
I think what needs to be sent over from verifier to attester is the optional parameters (if any) for the selected evidence format. For example, SGX local attestation would require the verifier to send over its SGX target info to the attester. But for ECDSA evidence, there is no need of any optional parameter.

Custom claims that the attester can embed in its evidence can include items such as the public key of the asymmetric key pair that serves as its cryptographic identity. This key pair is TEE agnostic, and the public key will be embedded in any TEE-specific evidence that it generates.

Reply:
Custom inputs to generating evidences are collected in the array of claims. They are (name,value) pairs so they can be used to pass in anything from verifier to attester. SGX local attestation is an unique. Ideally, SGX target info should be passed in as a claim.

---
In reply to: [390027862](https://github.com/openenclave/openenclave/pull/2621#discussion_r390027862) [](ancestors = 390027862)